src/main/kotlin/org/rust/ide/console/RsConsoleFolding.kt

```diff
@@ -0,0 +1,60 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.console;
+
+import com.intellij.execution.ConsoleFolding
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.text.StringUtil.pluralize
+import com.intellij.openapiext.isUnitTestMode
+import org.rust.cargo.project.model.cargoProjects
+import org.rust.cargo.project.workspace.PackageOrigin
+import org.rust.cargo.runconfig.filters.*
+import org.rust.lang.core.resolve.resolveStringPath
+import java.io.File
+import kotlin.math.max
+
+class RsConsoleFolding : ConsoleFolding() {
+    override fun getPlaceholderText(project: Project, lines: List<String>): String? {
+        if (lines.size < 2) return null
+
+        // each call has two lines - function name and source code location
+        val count = lines.size / 2
+        val callText = pluralize("call", count)
+        return "<$count internal $callText>"
+    }
+
+    override fun shouldFoldLine(project: Project, line: String): Boolean {
+        val backtraceFilters = project.cargoProjects.allProjects.map {
+            RsBacktraceItemFilter(project, it.workspace)
+        }
+        if (backtraceFilters.any {
+                val functionName = it.parseBacktraceRecord(line)?.functionName ?: return@any false
+                val func = FilterUtils.normalizeFunctionPath(functionName)
+                val workspace = it.workspace ?: return@any false
+                val (_, pkg) = resolveStringPath(func, workspace, project) ?: return@any true
+
+                pkg.origin != PackageOrigin.WORKSPACE
+            }) {
+            return true
+        }
+
+        val locationFilters = project.cargoProjects.allProjects.mapNotNull {
+            val dir = it.workspaceRootDir ?: it.rootDir
+            RegexpFileLinkFilter(project, dir ?: return@mapNotNull null, RsBacktraceFilter.LINE_REGEX)
+        }
+        return locationFilters.any {
+            val fileName = it.matchLine(line)?.groups?.get(1)?.value ?: return@any false
+            var platformNeutralName = fileName.replace(File.separatorChar, '/')
+            if (isUnitTestMode) {
+                // rootDir in unit tests points to a wrong directory
+                platformNeutralName = platformNeutralName.removePrefix("src/")
+            }
+            it.cargoProjectDirectory.findFileByRelativePath(platformNeutralName) == null
```

Comment:
Don't like that we made `cargoProjectDirectory` public only for this line.
You already have `dir` value in `mapNotNull`.
You can collapse `mapNotNull` and `any` calls into single `any` and use `dir`